core/src/main/java/org/infinispan/util/concurrent/locks/impl/DefaultLockManager.java

```diff
@@ -0,0 +1,395 @@
+      @Override
+      public void addListener(KeyAwareLockListener listener) {
+         lockPromise.addListener(state -> listener.onEvent(key, state));
+      }
+
+      @Override
+      public void run() {
+         lockPromise.cancel(LockState.TIMED_OUT);
+      }
+
+      public KeyAwareExtendedLockPromise scheduleLockTimeoutTask(ScheduledExecutorService executorService) {
+         if (executorService != null && timeoutMillis > 0 && !isAvailable()) {
+            ScheduledFuture<?> future = executorService.schedule(this, timeoutMillis, TimeUnit.MILLISECONDS);
+            lockPromise.addListener((state -> future.cancel(false)));
+         }
+         return this;
+      }
+   }
+
+   private static class CompositeLockPromise implements KeyAwareLockPromise, LockListener, Runnable {
+
+      private final List<KeyAwareExtendedLockPromise> lockPromiseList;
+      private final CompletableFuture<Void> notifier;
+      volatile LockState lockState = LockState.AVAILABLE;
+
+      private CompositeLockPromise(int size) {
+         lockPromiseList = new ArrayList<>(size);
+         notifier = new CompletableFuture<>();
+      }
+
+      public void addLock(KeyAwareExtendedLockPromise lockPromise) {
+         lockPromiseList.add(lockPromise);
+      }
+
+      public void markListAsFinal() {
+         for (LockPromise lockPromise : lockPromiseList) {
+            lockPromise.addListener(this);
+         }
+      }
+
+      @Override
+      public boolean isAvailable() {
+         for (LockPromise lockPromise : lockPromiseList) {
+            if (!lockPromise.isAvailable()) {
+               return false;
+            }
+         }
+         return true;
+      }
+
+      @Override
+      public void lock() throws InterruptedException, TimeoutException {
+         Exception exception = null;
+         ExceptionType exceptionType = ExceptionType.NONE;
+         for (ExtendedLockPromise lockPromise : lockPromiseList) {
+            try {
+               //we still need to invoke lock in all the locks.
+               lockPromise.lock();
+            } catch (InterruptedException e) {
+               if (exception == null) {
+                  exception = e;
+                  exceptionType = ExceptionType.INTERRUPTED;
+               }
+            } catch (TimeoutException e) {
+               if (exception == null) {
+                  exception = e;
+                  exceptionType = ExceptionType.TIMEOUT;
+               }
+            } catch (DeadlockDetectedException e) {
+               if (exception == null) {
+                  exception = e;
+                  exceptionType = ExceptionType.DEADLOCK;
+               }
+            } catch (RuntimeException e) {
+               if (exception == null) {
+                  exception = e;
+                  exceptionType = ExceptionType.RUNTIME;
+               }
+            }
+         }
+         if (exception != null) {
+            switch (exceptionType) {
+               case INTERRUPTED:
```

Comment:
I'd say if we get an `InterruptedException` we need to throw that, and/or re-set the thread's interruption status, regardless of any other exceptions. I had in mind something like this:

         InterruptedException interruptedException = null;
         TimeoutException timeoutException = null;
         DeadlockDetectedException deadlockException = null;
         RuntimeException runtimeException = null;
         for (ExtendedLockPromise lockPromise : lockPromiseList) {
            try {
               //we still need to invoke lock in all the locks.
               lockPromise.lock();
            } catch (InterruptedException e) {
               interruptedException = e;
            } catch (TimeoutException e) {
               timeoutException = e;
            } catch (DeadlockDetectedException e) {
               deadlockException = e;
            } catch (RuntimeException e) {
               runtimeException = e;
            }
         }
         if (interruptedException != null) {
            throw interruptedException;
         } else if (timeoutException != null) {
            throw timeoutException;
         } else if (deadlockException != null) {
            throw deadlockException;
         } else if (runtimeException != null) {
            throw runtimeException;
         }